working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,437 @@
+  LongTypeExpression x4,
+  LongTypeExpression x5,
+) {
+  ... // Lots of stuff.
+}
+```
+
+That proposal may certainly be helpful in the case where the primary
+constructor receives a large number of arguments with long types, etc.
+However, the proposal has not been included here. One reason is that it
+could be better to use a non-primary constructor whenever there is so much
+text. Also, it could be helpful to be able to search for the named
+constructor using `D.named`, and that would fail if we use the approach
+where it occurs as `new.named` or `const.named` because that particular
+constructor has been expressed as a primary constructor.
+
+A variant of this idea, from Leaf, was that we could allow one constructor
+in a class with no primary constructor in the header to be marked as a
+"primary constructor in the body". This would allow the constructor to have
+a body and an initializer list. As a strawman, let's say that we do this by
+adding the reserved word `var` in front of a normal constructor
+declaration:
+
+```dart
+class D<TypeVariable extends Bound> extends A with M implements B, C {
+  int i; 
+  
+  var D.named(
+    LongTypeExpression x1,
+    LongTypeExpression x2,
+    LongTypeExpression x3,
+    LongTypeExpression x4,
+    LongTypeExpression x5,
+  ) : 
+      i = 1, 
+      assert(x1 != x2), 
+      super.name(x3, y: x4) {
+    ... // Normal constructor body.
+  }
+
+  ... // Lots of stuff.
+}
+```
+
+The only special thing about a `var` constructor is that the non-super,
+non-this parameters are subject to the same processing as in a primary
+constructor, that is, they will introduce an instance variable. This
+proposal does not include that feature, but it should be completely
+compatible with the feature if we wish to add it.
+
+## Specification
+
+### Syntax
+
+The grammar is modified as follows. Note that the changes include
+support for inline classes, because they're intended to use primary
+constructors as well.
+
+```
+ <topLevelDefinition> ::=
+     <classDeclaration>
+   | <inlineClassDeclaration> // New alternative.
+   | ...;
+
+<classDeclaration> ::= // First alternative modified.
+     (<classModifiers> | <mixinClassModifiers>)
+     'class' <classNamePart> <superclass>? <interfaces>? <classBody>
+   | ...;
+
+<classNamePart> ::= // New rule.
+     'const'? <constructorName> <typeParameters>? <formalParameterList>
+   | <typeWithParameters>;
+
+<classBody> ::= // New rule.
+     '{' (<metadata> <classMemberDeclaration>)* '}'
+   | ';';
+
+<inlineClassDeclaration> ::=
+     'final'? 'inline' 'class' <classNamePart> <interfaces>? <inlineClassBody>;
+
+<inlineClassBody> ::=
+     '{' (<metadata> <inlineMemberDeclaration>)* '}'
+   | ';';
+
+<inlineMemberDeclaration> ::= <classMemberDeclaration>;
+```
+
+The word `inline` is now used in the grammar, but it is not a reserved word
+or a built-in identifier.
+
+### Static processing
+
+A class declaration (of any kind including `inline`) with a primary
```

Comment:
Except mixin application class declarations?

They can't declare instance variables anyway, and get forwarded constructors from the mixin application.
But maybe this can be a way to declare a different constructor, which can only have `super.x` parameters,          and which might suppress adding forwarding constructors?

Reply:
Let's just keep them out for now. That's already the case because the grammar rules don't include them.